src/query/procedure/module.cpp

```diff
@@ -1027,37 +1022,33 @@ bool PythonModule::Close() {
   py::Object sys(PyImport_ImportModule("sys"));
   PyObject *sys_mod_ref = sys.GetAttr("modules").Ptr();
 
-  std::filesystem::path submodules_path = file_path_.parent_path();
-  std::string_view stem = std::string_view(file_path_.stem().c_str());
-  submodules_path /= FLAGS_python_submodules_directory;
+  std::string stem = file_path_.stem().string();
 
   ProcessFileDependencies(file_path_, file_path_.stem().c_str(), func_code, sys_mod_ref);
 
-  if (std::filesystem::exists(submodules_path)) {
-    std::filesystem::path submodules;
+  std::filesystem::path submodules;
 
-    for (auto const &dir_entry : std::filesystem::directory_iterator(submodules_path)) {
-      std::string_view dir_entry_stem = std::string_view(dir_entry.path().stem().c_str());
-      if (dir_entry.is_regular_file() || dir_entry_stem.compare("__pycache__") == 0) continue;
-      if (dir_entry_stem.find(stem) != std::string_view::npos &&
-          (submodules.empty() || std::string_view(submodules.stem().c_str()).length() > dir_entry_stem.length())) {
-        submodules = dir_entry.path();
-      }
+  for (auto const &dir_entry : std::filesystem::recursive_directory_iterator(file_path_.parent_path())) {
+    std::string dir_entry_stem = dir_entry.path().stem().string();
+    if (dir_entry.is_regular_file() || dir_entry_stem == "__pycache__") continue;
+    if (dir_entry_stem.find(stem) != std::string_view::npos &&
+        (submodules.empty() || submodules.stem().string().size() > dir_entry_stem.length())) {
+      submodules = dir_entry.path();
```

Comment:
I think in case of `test_module.py` and `test_module_utils/test_module_internal` directories. This logic will choose `test_module_internal` as `submodules`.

I think a correct solution would be recursively iterating over the directory entries, and once we find a directory that contains `stem`, then
1. [disable the recursion](https://en.cppreference.com/w/cpp/filesystem/recursive_directory_iterator/disable_recursion_pending) on the iterator
2. Purge all `__pycache__` folders in this folder
3. Move on to the next one

Of course by introducing the recursion here we march into directories that are totally unnecessarily, but this shouldn't be called on the hot-path. Furthermore, it has the advantage that it works with mage without specifying any runtime flag.